netty/src/main/java/io/grpc/netty/AbstractNettyHandler.java

```diff
@@ -157,7 +174,12 @@ public void onDataRead(int dataLength, int paddingLength) {
       if (!autoTuneFlowControlOn) {
         return;
       }
-      if (!isPinging() && pingLimiter.isPingAllowed()) {
+      if (lastTargetWindow == 0 && pingCount > 0) {
```

Comment:
Does `pingCount > 0` just avoid this delaying the first ping? Could we instead move this condition after the `!isPinging()` condition and remove the `pingCount > 0`? That'd make it more obvious it happens after. Also, we send pings for reasons other than BDP, so `pingCount` is imprecise.

Reply:
Note that the old line 160 and the new line 182 are identical.  It is really just lines 178-181 being inserted before, so just a question of whether to leave lastTargetWindow as 0 to allow the first ping to go without the delay.  We want the other pings to be delayed if we haven't gotten enough data.

Reply:
This pingCount value is being stored in FlowControlPinger, so I believe that it is only tracking the BDP pings.

Reply:
Oh, `pingCount` was previously unused except for tests. I was thinking it was the counter that is actually saved in PingLimiter.

> so just a question of whether to leave lastTargetWindow as 0 to allow the first ping to go without the delay

I was fine with that part. But to achieve the same goal it seemed clearer to swap the two `if`s. (I think there was confusion when I said "condition" which could mean the entire if or each part of the if.)

```
if (!isPinging() && pingLimiter.isPingAllowed() ...) {
  ...
}
if (lastTargetWindow == 0) {
  ...
}
```

I can live with the `pingCount > 0`, but wanted to offer the alternative.

Reply:
After your comment, walking through the logic I confirmed that lastTargetWindow will never get reset to 0, so the new value for it isn't needed for the comparison in the if(!isPinging()... .  So, moving as you suggest.